### DISCUSSION TITLE
AZURE openid Invalid user credentials.

### DISCUSSION THREAD

**Discussion by Kminkjan** [directus/directus#10864] [Archive]
Hi, I'm trying to add Microsoft Azure AD SSO to my instance of Directus. While it seems to be connecting to the openid endopoint, it always returns an "Invalid user credentials." in the logs and a https://<mydirectusurl>/admin/login?reason=INVALID_USER error.

What I did discover is that, when I use `AUTH_MICROSOFT_ALLOW_PUBLIC_REGISTRATION="true"` in the .env file, the directus account ends up being created with all the right info and an identifier key, but I still get the "invalid user credentials" when logging in.

Now I'm not sure if this is a bug, or I still have something configured wrong at my azure. Does anyone have steps or ideas so I can debug this further?

The SSO params in my .env file:

```
## Auth Providers

AUTH_PROVIDERS="microsoft"
AUTH_MICROSOFT_DRIVER="openid"
AUTH_MICROSOFT_IDENTIFIER_KEY="sub"
AUTH_MICROSOFT_CLIENT_ID="<redacted>"
AUTH_MICROSOFT_CLIENT_SECRET="<redacted>"
AUTH_MICROSOFT_SCOPE="openid profile email"
AUTH_MICROSOFT_ALLOW_PUBLIC_REGISTRATION="true"
AUTH_MICROSOFT_ISSUER_URL="https://login.microsoftonline.com/<redacted>/v2.0/.well-known/openid-configuration"
```

**Comment by azrikahar** [comment]
Hi, can you try setting `email` instead of `sub` for identifier key? The identifier needs to be the same as the user's email in Directus (vice versa) for the login to be valid. Here's a working config shared by another community member here: https://github.com/directus/directus/discussions/10511#discussioncomment-1814764

**Reply by aidenfoxx** [reply]
Just to clarify, you can use other identifiers, but the identifier specified by `AUTH_<provider>_IDENTIFIER_KEY` must be appropriate for the users "External Identifier". So for example:

- If the users "External Identifier" is an email, you should use `AUTH_<provider>_IDENTIFIER_KEY="email"`
- If the users "External Identifier" is a phone number, you should use `AUTH_<provider>_IDENTIFIER_KEY="phone"`

The default `sub` is the users unique OpenID identifier, and is the optimal identifier for dynamic user registration.

**Reply by Kminkjan** [reply]
I tried to use the email AUTH_MICROSOFT_IDENTIFIER_KEY as well (making sure the email is also in the identifier key field in the directus user). And got the same error!

**Comment by aidenfoxx** [comment]
@Kminkjan I assume you're trying to log in with the users Email and Password? Users can only authenticate with the authentication "Provider" they were created with, so your user will only be able to authenticate via your Microsoft OpenID provider.

**Comment by Kminkjan** [comment]
Thanks for your reactions! Well I have created the user by using the "Log in with microsoft" button on the login screen:
![image](https://user-images.githubusercontent.com/2970115/148241214-00d2cf19-482b-4e9e-835b-63b469e197a4.png)

After that moment, when I log in with another (admin) account I can see that the user has been created with an identifier (I've cut the identifier short in the image):
![image](https://user-images.githubusercontent.com/2970115/148241454-2b9a0518-f08f-43d9-b4ca-76fac64caf2a.png)

So it seems to mee that Directus has bound this account to that microsoft openid. Right?

Then when I logout and click on "Log in with microsoft again", I get this:
![image](https://user-images.githubusercontent.com/2970115/148241820-bf0cbecd-08d2-4f59-bcad-4e71dc9117a9.png)

**Reply by aidenfoxx** [reply]
That's odd. Hard to know what's going on without any logs. Do you see errors or messages?

**Comment by Kminkjan** [comment]
Well I've found out the problem: when I gave the user a role, the SSO worked! Didn't know they explicitly needed roles (but makes sense).
Thing is, I cant seem to make `AUTH_MICROSOFT_DEFAULT_ROLE_ID="640994fe-1e39-4dcc-b514-215534a0bc1b"` work. `640994fe-1e39-4dcc-b514-215534a0bc1b` is the id of the role. Am I doing that wrong?

**Reply by aidenfoxx** [reply]
It won't be apply a role retroactively. Only for any newly created users.

**Reply by Kminkjan** [reply]
Oh yes, I was making new users, I should have made that clear.

But it turned out that I was a bit stupid and wrote `MIRCOSOFT` instead of `MICROSOFT` ... End of the workday oversight ;)

Thank you so much for your time!
